src/mame/machine/k573fpga.cpp

```diff
@@ -0,0 +1,592 @@
+        (IsBitSet(v, b6) << 6) |
+        (IsBitSet(v, b5) << 5) |
+        (IsBitSet(v, b4) << 4) |
+        (IsBitSet(v, b3) << 3) |
+        (IsBitSet(v, b2) << 2) |
+        (IsBitSet(v, b1) << 1) |
+        (IsBitSet(v, b0) << 0);
+}
+
+inline uint16_t k573fpga_device::fpga_decrypt_byte_real(uint16_t v)
+{
+    uint16_t m = crypto_key1 ^ crypto_key2;
+
+    v = BitSwap16(
+        v,
+        15 - IsBitSet(m, 0xF),
+        14 + IsBitSet(m, 0xF),
+        13 - IsBitSet(m, 0xE),
+        12 + IsBitSet(m, 0xE),
+        11 - IsBitSet(m, 0xB),
+        10 + IsBitSet(m, 0xB),
+        9 - IsBitSet(m, 0x9),
+        8 + IsBitSet(m, 0x9),
+        7 - IsBitSet(m, 0x8),
+        6 + IsBitSet(m, 0x8),
+        5 - IsBitSet(m, 0x5),
+        4 + IsBitSet(m, 0x5),
+        3 - IsBitSet(m, 0x3),
+        2 + IsBitSet(m, 0x3),
+        1 - IsBitSet(m, 0x2),
+        0 + IsBitSet(m, 0x2)
+    );
+
+    v ^= (IsBitSet(m, 0xD) << 14) ^
+        (IsBitSet(m, 0xC) << 12) ^
+        (IsBitSet(m, 0xA) << 10) ^
+        (IsBitSet(m, 0x7) << 8) ^
+        (IsBitSet(m, 0x6) << 6) ^
+        (IsBitSet(m, 0x4) << 4) ^
+        (IsBitSet(m, 0x1) << 2) ^
+        (IsBitSet(m, 0x0) << 0);
+
+    v ^= BitSwap16(
+        crypto_key3,
+        7, 0, 6, 1,
+        5, 2, 4, 3,
+        3, 4, 2, 5,
+        1, 6, 0, 7
+    );
+
+    return (v >> 8) | ((v & 0xff) << 8);
+}
+
+inline uint16_t k573fpga_device::fpga_decrypt_byte_fake(uint16_t data, uint32_t crypto_idx)
+{
+    // Not the real algorithm. Only used for DDR Solo Bass Mix (ddrsbm)
+    int32_t crypto_enc_idx = find_enc_key();
+
+    if (crypto_enc_idx == -1) {
+        return data;
+    }
+
+    uint8_t *key = k573enc_keys[crypto_enc_idx];
+
+    uint16_t output_word = 0;
+    for (int cur_bit = 0; cur_bit < 8; cur_bit++) {
+        int even_bit_shift = (cur_bit * 2) & 0xff;
+        int odd_bit_shift = (cur_bit * 2 + 1) & 0xff;
+        int is_even_bit_set = (data & (1 << even_bit_shift)) != 0;
+        int is_odd_bit_set = (data & (1 << odd_bit_shift)) != 0;
+        int is_key_bit_set = (key[crypto_idx % 16] & (1 << cur_bit)) != 0;
+        int is_scramble_bit_set = (key[(crypto_idx - 1) % 16] & (1 << cur_bit)) != 0;
+
+        if (is_scramble_bit_set == 1) {
+            int t = is_even_bit_set;
+            is_even_bit_set = is_odd_bit_set;
+            is_odd_bit_set = t;
+        }
+
+        if (((is_even_bit_set ^ is_key_bit_set)) == 1) {
+            output_word |= 1 << even_bit_shift;
+        }
+
+        if ((is_odd_bit_set) == 1) {
+            output_word |= 1 << odd_bit_shift;
+        }
+    }
+
+    return (output_word >> 8) | ((output_word & 0xff) << 8);
+}
+
+SAMPLES_UPDATE_CB_MEMBER(k573fpga_device::k573fpga_stream_update)
+{
+    if ((mpeg_ctrl_flag & 0xe000) != 0xe000 || mp3_last_adr >= mp3_end_adr || (m_samples->get_position(0) < mp3_next_sync && mp3_next_sync != 0)) {
+        return;
+    }
+
+    int new_samples = 0;
+
+    if (!use_fake_fpga && mp3_start_adr != mp3_dynamic_base) {
```

Comment:
Eventually that will be fully streamed and most of that code will just go away, but that will require a working mpeg_audio.cpp.